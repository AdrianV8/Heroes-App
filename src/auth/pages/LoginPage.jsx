import { useContext } from "react";
import { useNavigate } from "react-router-dom";
import { AuthContext } from "../context";

export const LoginPage = () => {

  const { login } = useContext(AuthContext)
  const navigate = useNavigate();

  const onLogin = () => {

    const lastPath = localStorage.getItem('lastPath') || '/';
  
    login('Adrian Lopera')


    navigate(lastPath ,{
      replace: true
    })
  }
  
  return (
    <>
      <div className="container mt-5 animate__animated animate__fadeInLeft animate__delay-0.5s">
        <h1>Login</h1>
        <hr />

        <button 
          className="btn btn-success"
          onClick={ onLogin }
          >
          Login
        </button>
        
      </div>
    </>
  )
}
