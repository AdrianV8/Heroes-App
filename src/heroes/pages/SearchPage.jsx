import { useLocation, useNavigate } from "react-router-dom";
import queryString from "query-string";
import { useForm } from "../hooks";
import { HeroSearchResults } from "../components/HeroSearchResults";

export const SearchPage = () => {
  
  // useNavigate: obtener la navegación
  const navigate = useNavigate();
  // useLocation: obtener la información de la localización actual
  const location = useLocation();
  /**
   * ----¡OPCIONAL!----
   *  query-string sirve para ver mejor los parámetros que se envian
   *  Las propiedades del objeto location: search -> Son los parámetros del query parameters(?q='Batman'&asc=true)
   *  Log para ver todas las propiedades del objeto 'location'
   *  El parámetro q, es el nombre que buscamos en el formulario
   */
  const { q='' } = queryString.parse(location.search);

  
  // Implementar useForm y rescatar el valor del form
  const { searchText, onInputChange } = useForm({
    searchText: "",
  });

  const onSearchSubmit = (event) => {
    event.preventDefault();
    // if (searchText.trim().length <= 1) return;

    // Enviar query parameters
    navigate(`?q=${searchText.toLowerCase().trim()}`);
  };

  return (
    <>
      <h1>Search</h1>
      <hr />
      <div className="row">
        <div className="col-5">
          <h4>Búsquedas</h4>
          <hr />

          <form onSubmit={onSearchSubmit}>
            <input
              type="text"
              placeholder="Buscar un Héroe"
              className="form-control"
              name="searchText"
              autoComplete="off"
              value={searchText}
              onChange={onInputChange}
            />
            <button className="btn btn-outline-primary mt-1">Buscar</button>
          </form>
        </div>

        <div className="col-7">
          <h4>Resultados</h4>
          <hr />

          <HeroSearchResults queryParameter={q}/>

        </div>
      </div>
    </>
  );
};
