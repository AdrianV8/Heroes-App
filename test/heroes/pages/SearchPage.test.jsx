import { fireEvent, render, screen } from "@testing-library/react";
import { MemoryRouter, useNavigate } from "react-router-dom";

import { SearchPage } from "../../../src/heroes/pages/SearchPage";

const mockNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockNavigate
}))

describe("Pruebas Search Page", () => {

    beforeAll( () => jest.clearAllMocks());

    test("Debe de mostrarse los valores por defecto", () => {

        const { container } = render(
            <MemoryRouter>

                <SearchPage />

            </MemoryRouter>
        
        );
        expect( container ).toMatchSnapshot();
  
    });

    test("Debe de mostrar el héroe y el input con el valor del queryString", () => {

        render(
            <MemoryRouter initialEntries={['/search?q=batman']}>

                <SearchPage />

            </MemoryRouter>
        
        );

        const img = screen.getByRole('img');
        expect( img.src ).toContain('/assets/heroes/dc-batman.jpg')
  
    });
    test("Debe de mostrar el héroe y mantener oculto mensajes de error y búsqueda", () => {

        render(
            <MemoryRouter initialEntries={['/search?q=batman']}>

                <SearchPage />

            </MemoryRouter>
        
        );

        const img = screen.getByRole('img');
        expect( img.src ).toContain('/assets/heroes/dc-batman.jpg')

        const searchHero = screen.getByLabelText('search_hero');
        const searchError = screen.getByLabelText('search_error');
        expect( searchHero.style.display ).toBe('none');
        expect( searchError.style.display ).toBe('none');
  
    });

    test("Debe mostrar un error si no se encuentra el heroe", () => {

        render(
            <MemoryRouter initialEntries={['/search?q=batman123']}>

                <SearchPage />

            </MemoryRouter>
        
        );

        const searchError = screen.getByLabelText('search_error');
        expect( searchError.style.display ).toBe('');
  
    });
  
    test("Debe de llamar en navigate a la pantalla nueva", () => {

        const inputValue = 'superman';

        render(
            <MemoryRouter initialEntries={['/search']}>

                <SearchPage />

            </MemoryRouter>
        
        );

        const input = screen.getByRole('textbox');
        const form = screen.getByLabelText('form');

        fireEvent.input(input, {target: {value: inputValue}});
        fireEvent.submit(form);
        expect(mockNavigate).toHaveBeenCalledWith(`?q=${inputValue}`)
  
    });

    
});
