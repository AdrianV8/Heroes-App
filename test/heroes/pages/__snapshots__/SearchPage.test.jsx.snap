// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas Search Page Debe de mostrarse los valores por defecto 1`] = `
<div>
  <h1>
    Search
  </h1>
  <hr />
  <div
    class="row"
  >
    <div
      class="col-5"
    >
      <h4>
        Búsquedas
      </h4>
      <hr />
      <form
        aria-label="form"
      >
        <input
          autocomplete="off"
          class="form-control"
          name="searchText"
          placeholder="Buscar un Héroe"
          type="text"
          value=""
        />
        <button
          class="btn btn-outline-primary mt-1"
        >
          Buscar
        </button>
      </form>
    </div>
    <div
      class="col-7"
    >
      <h4>
        Resultados
      </h4>
      <hr />
      <div
        aria-label="search_hero"
        class="alert alert-primary text-center animate__animated animate__fadeIn"
      >
        Busca un héroe
      </div>
      <div
        aria-label="search_error"
        class="alert alert-danger animate__animated animate__fadeIn"
        style="display: none;"
      >
        No se ha encontrado resultados de 
        <b />
      </div>
    </div>
  </div>
</div>
`;
